Add user guide for data transforms

JakeSCahill opened the pull request:
## Description

Resolves https://github.com/redpanda-data/documentation-private/issues/2242
Resolves https://github.com/redpanda-data/documentation-private/issues/2341
Resolves https://github.com/redpanda-data/documentation-private/issues/2572
Review deadline: 1 July

## Page previews

[Data transforms docs preview](https://deploy-preview-567--redpanda-docs-preview.netlify.app/current/develop/data-transforms/)

Related PRs: https://github.com/redpanda-data/docs-ui/pull/205 and https://github.com/redpanda-data/docs-extensions-and-macros/pull/56

## Checks

- [ ] New feature
- [x] Content gap
- [ ] Support Follow-up
- [ ] Small fix (typos, links, copyedits, etc)

oleiman commented:
cc @rockwotj!

rockwotj commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -0,0 +1,655 @@
+
+When you deploy this transform function, and produce a message without a key, you'll get the following in the logs:
+
+[source,js,role="no-copy"]
+----
+{
+  "body": {
+    "stringValue": "2024/06/20 08:17:33 Error: Record key is nil\n"
+  },
+  "timeUnixNano": 1718871455235337000,
+  "severityNumber": 13,
+  "attributes": [
+    {
+      "key": "transform_name",
+      "value": {
+        "stringValue": "test"
+      }
+    },
+    {
+      "key": "node",
+      "value": {
+        "intValue": 0
+      }
+    }
+  ]
+}
+----
```
This is a good opportunity to showcase `rpk transform logs`, because we don't tell people how to see this. Using the rpk command we can show people the command that generates this.

rockwotj commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -0,0 +1,655 @@
+
+See also:
+
+- xref:manage:schema-reg/schema-reg-overview.adoc[].
```
Should we be consistent with `.`?

rockwotj commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -0,0 +1,655 @@
+  client := sr.NewClient()
+	transform.OnRecordWritten(doNothing)
+
+  // Use the client...
```
You can't use the client here, `transform.OnRecordWritten` runs forever and never returns.

This should be passed into the callback, or registered as a global variable (same for Rust).

rockwotj commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -0,0 +1,655 @@
+--
+======
+
+==== Read schemas from the Schema Registry
```
This is all awesome, but we should show people how to interact with schemas when transforms are being executed as that's generally what people would want. Here's the pattern I expect in psuedo code:

```
client: sr_client

my_schema = "..."
my_schema_id = -1

main()
  client = create_sr_client()
  my_schema_id = write_schema(my_schema)
  on_record_written(my_transform)

my_transform(record, writer)
  schema_id, value = get_schema_id(record.value())
  record_schema = sr_client.get_schema(schema_id)
  struct = deserialize(value, my_schema, record_schema) # Assuming compatibility
  new_struct = do_transform(struct)
  new_value = serialize(new_struct, my_schema)
  new_value = add_schema_id(new_value, my_schema_id)
  writer.write(Record(record.key(), new_value))
```

we can certainly make this experience better over time.

oleiman replied:
worth noting that rust is not particularly great at "global instance of complicated (possibly) mutable thing". 🤷

JakeSCahill replied:
thanks! @rockwotj or @oleiman do you have a simple working example of this in Go and Rust?

For Go, something like

```go
package main

import (
  "context"
  "fmt"
  "github.com/redpanda-data/redpanda/src/transform-sdk/go/transform"
  "github.com/redpanda-data/redpanda/src/transform-sdk/go/transform/sr"
)

func main() {
  // Initialize the Schema Registry client
  client := sr.NewClient()

  // Register your transform function.
  transform.OnRecordWritten(doTransform)
}

func doTransform(event transform.WriteEvent, writer transform.RecordWriter) error {
  // Lookup schema by ID
  schema, err := client.LookupSchemaById(1)
  if err != nil {
    panic(err)
  }

  fmt.Println("Schema:", schema.Schema)
  return writer.Write(event.Record())
}
```

I'm not sure how to get a reference to the global client.

oleiman replied:
>working examples...?

Qualified yes:
- [golang](https://github.com/redpanda-data/redpanda/blob/dev/src/transform-sdk/go/transform/internal/testdata/schema-registry/transform.go) 
- [rust](https://github.com/redpanda-data/redpanda/blob/dev/src/transform-sdk/rust/examples/schema_registry/main.rs)

Qualified in the sense that these were written for integration testing and might not be suitable as docs-facing examples w/o some modification. Each exhibits a possible pattern for initializing and consuming an SR client though.

JakeSCahill replied:
Let's leave the examples for now and before we release 24.2 we can update them to show how to use JSON schemas.

JakeSCahill replied:
Since that's a new feature, we might want to showcase it here

rockwotj commented on modules/develop/pages/data-transforms/configure.adoc:
```diff
@@ -0,0 +1,116 @@
+
+Transform functions are isolated from the broker's internal environment variables to maintain security and encapsulation. Each transform function only uses the environment variables explicitly provided to it.
+
+NOTE: Built-in environment variables cannot be overwritten. Attempting to overwrite them in your custom environment variables results in an `Invalid transform environment` error during deployment.
```
We actually reserve anything that starts with `REDPANDA_` - Also there are size restrictions: https://github.com/redpanda-data/redpanda/blob/98e48acfb4dc7140328cc0c109a88f333f08a92f/src/v/cluster/plugin_frontend.cc#L282-L374

Key must be under 128 bytes, value under 2K total of 128 key value pairs, must be valid UTF8, no control characters (null bytes etc).

JakeSCahill replied:
> total of 128 key value pairs

Is this including the built-in ones?

rockwotj replied:
No just user defined

rockwotj commented on modules/develop/pages/data-transforms/deploy.adoc:
```diff
@@ -0,0 +1,95 @@
++
+[source,bash]
+----
+rpk transform deploy --file=<path-to-wasm-binary> --var <environment-variable-name>=<value>
```
You don't need the `file` flag or the `var` flag if you're using transform.yaml

I recommend everyone use a `transform.yaml` file, but if you want to override a value on the CLI, you can do that (they get merged), or you can do all flags and drop the config file.

rockwotj commented on modules/develop/pages/data-transforms/deploy.adoc:
```diff
@@ -0,0 +1,95 @@
+
+. xref:develop:data-transforms/build.adoc[Make you changes to the code].
+. <<build, Rebuild>> the Wasm binary.
+. <<deploy, Redeploy>> the Wasm binary to the same Redpanda cluster.
```
Maybe useful to note that if you supplied values on the command line those can be overwritten the next deploy but they aren't "remembered" per se

rockwotj commented on modules/develop/pages/data-transforms/test.adoc:
```diff
@@ -0,0 +1,83 @@
+= Write Integration Tests for Transform Functions
```
This is awesome ❤️

rockwotj commented on modules/develop/pages/data-transforms/test.adoc:
```diff
@@ -0,0 +1,83 @@
+
+Integration tests verify that your transform functions work correctly in a real Redpanda environment. You can use https://github.com/testcontainers/testcontainers-go/tree/main[testcontainers] to set up and manage a Redpanda instance for testing.
+
+For more detailed examples and helper code for setting up integration tests, refer to the SDK integration tests on https://github.com/redpanda-data/redpanda/tree/dev/src/transform-sdk/tests[GitHub].
```
We should work to moving this to a better example in our labs repo (no need to do anything right now). Mark shared some good stuff in community slack on this.

JakeSCahill replied:
https://github.com/redpanda-data/documentation-private/issues/2568

rockwotj commented on modules/develop/pages/data-transforms/monitor.adoc:
```diff
@@ -0,0 +1,70 @@
+= Monitor Data Transforms
```
👏

rockwotj commented on modules/develop/pages/data-transforms/versioning-compatibility.adoc:
```diff
@@ -0,0 +1,67 @@
+
+== Backward compatibility
+
+SDK versions will continue to work in the supported versions of Redpanda. However, they may not work in newer versions of Redpanda. Always check the compatibility matrix when upgrading either the SDK or Redpanda.
```
> However, they may not work in newer versions of Redpanda.

These two sentences to me counter each other.

Maybe something like:

SDK versions will have backwards compatibilty and continuted to be supported in newer versions of the broker. However, newer SDK versions may not work with older versions of Redpanda. Always check this compatibility matrix when upgrading the SDK or Redpanda.

oleiman replied:
In general, directing users to the compatibility matrix is probably best, but I think the common case will be for new versions of redpanda to maintain compatibility with older SDK versions. Otherwise a transform might die across a broker upgrade?

Whereas new SDK versions may require a newer redpanda version to function.

rockwotj reviewed (commented):
<img src="https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExbXUxcmN2NnpweWhscnJrMHFmNDZpb3V3ZWJmbThmc3RoYmJvamFyNCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/dmM5DoQTTElG7uzWiE/giphy.gif">

oleiman reviewed (commented):
So awesome. Thanks for all this ❤️

rockwotj commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -0,0 +1,654 @@
+
+You can develop your transform logic with one of the available xref:reference:data-transforms/sdks.adoc[software development kits (SDKs)] that allow you to run data transforms on a Redpanda cluster.
+
+All transform functions must register a callback with the `OnRecordWritten()` method.
```
PS unless you snuck in a PR I missed JS will have registration too with `onRecordWritten`

oleiman replied:
no, i am just javascript blind. I did not notice that the generated identity transform is passing a lambda into it.

JakeSCahill replied:
I'll add docs for JS support after this PR is merged.

JakeSCahill commented on modules/develop/pages/data-transforms/deploy.adoc:
```diff
@@ -0,0 +1,95 @@
+
+=== Invalid transform environment
+
+This error means that one or more custom environment variables have the same name as the xref:develop:data-transforms/configure.adoc#built-in[built-in ones].
+
+Change the name of those custom environment variables to something unique.
```
Given this, I'm going to add the limitations to a central place and link to it from here

https://github.com/redpanda-data/docs/pull/567#discussion_r1660391784

JakeSCahill commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -0,0 +1,654 @@
+--
+[source,go]
+----
+package main
+
+import (
+	"github.com/redpanda-data/redpanda/src/transform-sdk/go/transform"
+)
+
+func main() {
+	transform.OnRecordWritten(handleRecord)
+}
+
+// Required callback, even if it does nothing
+func handleRecord(event transform.WriteEvent, writer transform.RecordWriter) error {
+	return nil
+}
```
It's strange that the boilerplate function names differ between Rust and Go. We should update that in `rpk`:

```go
package main

import (
	"github.com/redpanda-data/redpanda/src/transform-sdk/go/transform"
)

func main() {
        // Register your transform function. 
        // This is a good place to perform other setup too.
	transform.OnRecordWritten(doTransform)
}

// doTransform is where you read the record that was written, and then you can
// output new records that will be written to the destination topic
func doTransform(e transform.WriteEvent, w transform.RecordWriter) error {
	return w.Write(e.Record())
}
```

```rust
use redpanda_transform_sdk::*;

fn main() {
  // Register your transform function.
  // This is a good place to perform other setup too.
  on_record_written(my_transform);
}

// my_transform is where you read the record that was written, and then you can
// return new records that will be written to the output topic
fn my_transform(event: WriteEvent, writer: &mut RecordWriter) -> Result<(), Box<dyn Error>> {
  writer.write(event.record)?;
  Ok(())
}
```

JakeSCahill replied:
I'd rather the arguments be written in full like in the Rust example too.

rockwotj replied:
We should name them both my transform.

As for the argument naming, that is sort of a go-ism, but we can consult more gophers on the team and get their thoughts.

https://go.dev/wiki/CodeReviewComments#variable-names

JakeSCahill replied:
TIL 😄 thanks!

JakeSCahill commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -0,0 +1,654 @@
+--
+======
+
+When creating a transform function, you should run any initialization steps in the `main()` function because it's only run once when the transform function is first deployed. If you're using Go, you can also use the standard predefined `init()` function.
```
Adding a link to https://golangdocs.com/init-function-in-golang

JakeSCahill commented on modules/develop/pages/data-transforms/configure.adoc:
```diff
@@ -0,0 +1,116 @@
+
+NOTE: Built-in environment variables cannot be overwritten. Attempting to overwrite them in your custom environment variables results in an `Invalid transform environment` error during deployment.
+
+== Configure the Wasm engine
```
Those are documented in the linked reference docs.

JakeSCahill commented on modules/develop/pages/data-transforms/versioning-compatibility.adoc:
```diff
@@ -0,0 +1,67 @@
+|===
+| Golang SDK Version | Redpanda Version
+
+| 1.0.x       | 24.1.x, 23.3.x
+| 0.2.x       | 23.3.x
+| 0.1.x       | 23.3.x
```
Good point. I'll remove those prerelease versions.

JakeSCahill commented:
Thanks, both 😄 I've addressed all your feedback in https://github.com/redpanda-data/docs/pull/567/commits/fdc71a0f84b4d5db945becd3063a0fd38a2303cb

I'm unsure about the Schema Registry code examples. If you have suggestions for those, please could you provide the code?

JakeSCahill commented on modules/develop/pages/data-transforms/test.adoc:
```diff
@@ -0,0 +1,83 @@
+= Write Integration Tests for Transform Functions
```
Added to https://github.com/redpanda-data/documentation-private/issues/2568

rockwotj commented:
I can write something up when I am back from OOO, but would not hold up all this great work before them.

oleiman commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -121,25 +199,19 @@ import (
-  // Ensure no error is returned
-  return nil
+  return writer.Write(record)
 }
```
idle thought: broadly this seems feels like a complete example, exhibiting both error-handling behaviors. If we detect a "bad" record, it can be discarded outright and processing continues. If writing a "good" record _fails_, we allow that error to propagate.

JakeSCahill replied:
Thanks! I've cleaned this part up to include this single complete example.

oleiman commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -338,14 +407,14 @@ import (
 
 func main() {
   client := sr.NewClient()
```
Think we want the client to be global here, like this:

https://github.com/redpanda-data/redpanda/blob/c7e27f720856b1787e4b26164863c3f7597acc1f/src/transform-sdk/go/transform/internal/testdata/schema-registry/transform.go#L26-L33

oleiman commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -358,15 +427,15 @@ use redpanda_transform_sdk::*;
-// Required callback, even if it does nothing
-fn do_nothing(_event: WriteEvent, _writer: &mut RecordWriter) -> anyhow::Result<()> {
+fn my_transform(_event: WriteEvent, _writer: &mut RecordWriter) -> anyhow::Result<()> {
+  // Use the client
```
>// Use the client

Can't do a straight global like go. I'm sure there's a subjectively cleaner way to accomplish this, but here's what we do for integration testing:

https://github.com/redpanda-data/redpanda/blob/c7e27f720856b1787e4b26164863c3f7597acc1f/src/transform-sdk/rust/examples/schema_registry/main.rs#L31-L47

oleiman commented on modules/develop/pages/data-transforms/build.adoc:
```diff
@@ -391,27 +460,27 @@ package main
 
 func main() {
   // Initialize the Schema Registry client
   client := sr.NewClient()
```
same comment. this scope doesn't extend to the transform function body.

oleiman commented on modules/develop/pages/data-transforms/deploy.adoc:
```diff
@@ -65,19 +63,21 @@ rpk transform deploy --file=https://my-site/my-transform.wasm
 . <<build, Rebuild>> the Wasm binary.
 . <<deploy, Redeploy>> the Wasm binary to the same Redpanda cluster.
++
+If you provided configuration flags on the command line, those can be overwritten when you redeploy the Wasm binary.
```
nitpick: more strict

```suggestion
If needed, deploy-time configuration overrides must be provided every time you deploy a Wasm binary. Otherwise they will be overwritten by default values and/or the contents of `transform.yaml`.
```

or some such. that's sort of clumsy.

oleiman reviewed (commented):
Looking really good. Most of my comments are around sample code.

I'm wondering whether we should hold off on the SR examples for now (put "examples coming soon" or something) until we can dial something in w/ Tyler. There's a bit of lang-specific mess to sort through, and it might be easier for everyone if we can standardize on something robust.

JakeSCahill commented:
Thanks @oleiman - in the interest of getting this merged today, I've removed the Schema Registry examples for now.

oleiman reviewed (approved):
lgtm 

:shipit:

kbatuigas commented on modules/develop/pages/data-transforms/how-transforms-work.adoc:
```diff
@@ -42,10 +44,31 @@ A new transform function reads the input topic from the latest offset. That is,
+- *Single record transforms*: Only single record transforms are supported, but multiple output records from a single input record are supported. For aggregations, joins, or complex transformations, consider xref:redpanda-connect:ROOT:about.adoc[Redpanda Connect] or https://flink.apache.org/[Apache Flink].
+- *Output topic limit*: Up to eight output topics are supported.
+- *Delivery semantics*: Transform functions have at-least-once delivery.
+- *Transaction API*: When clients use the Kafka Transactions API on partitions of an input topic, transform functions process only committed records.
```
```suggestion
- *Transactions API*: When clients use the Kafka Transactions API on partitions of an input topic, transform functions process only committed records.
```
nit

kbatuigas commented on modules/develop/pages/data-transforms/how-transforms-work.adoc:
```diff
@@ -42,10 +44,31 @@ A new transform function reads the input topic from the latest offset. That is,
+=== Environment variable limitations
+
+- *Maximum number of variables*: You can set up to 128 custom environment variables.
+- *Reserved prefix*: Variable keys must not start with `REDPANDA_`. This prefix is reserved for <<built-in,built-in environment variables>>.
```
```suggestion
- *Reserved prefix*: Variable keys must not start with `REDPANDA_`. This prefix is reserved for xref:develop:data-transforms/configure.adoc#built-in[built-in environment variables].
```
Moved to a new page?

kbatuigas commented on modules/develop/pages/data-transforms/versioning-compatibility.adoc:
```diff
@@ -0,0 +1,60 @@
+| 1.0.x       | 24.1.x, 23.3.x
+|===
+
+== Best Practices
```
```suggestion
== Best practices
```

kbatuigas reviewed (changes requested):
This all looks really good, bravo @JakeSCahill ! Just requesting a couple of minor changes.

The only other question I had was whether we have indeed decided to move away from publishing the reference docs on docs.redpanda.com. I know we've started to help review the reference content in the source PRs anyway, but curious to know if it was a product decision to publish the reference in just the package/crate docs (which to me implies that the docs team won't be maintaining it).

JakeSCahill commented:
Thanks for catching those @kbatuigas! I've committed all your feedback.

For the SDK reference, we agreed that since the SDK docs are autogenerated for Go and Rust, we should use those as the source of truth. It prevents APIs from becoming outdated in our docs. For example, the Rust SDK reference in our docs didn't include any Schema Registry info, but the autogenerated crate docs did. https://redpandadata.slack.com/archives/C01819S3ARH/p1719249225254979

JakeSCahill commented:
@oleiman @kbatuigas place could you take another quick look when you're back Friday? I changed the base to the beta branch since we need to get this one in: https://github.com/redpanda-data/documentation-private/issues/2572

https://github.com/redpanda-data/docs/pull/567/commits/1e1c197aeb2200e503d04da633a1d0bf640b53e1

kbatuigas commented on modules/develop/pages/data-transforms/deploy.adoc:
```diff
@@ -0,0 +1,133 @@
+
+You can also deploy data transforms on a Redpanda cluster by providing an addressable path to the Wasm binary. This is useful for sharing transform functions across multiple clusters or teams within your organization.
+
+For example, if the Wasm binary is hosted at `https://my-site/my-transform.wasm`, use the following command to deploy it:
```
The URL renders as a clickable link, we could consider changing the formatting so that it's just a formatted string.

JakeSCahill replied:
```suggestion
For example, if the Wasm binary is hosted at `\https://my-site/my-transform.wasm`, use the following command to deploy it:
```

kbatuigas reviewed (approved):
🚀

oleiman commented on modules/develop/pages/data-transforms/deploy.adoc:
```diff
@@ -48,6 +47,43 @@ You can deploy your transform function using the xref:reference:rpk/rpk-transfor
+
+In some cases, you may need to reprocess records from an input topic that already contains data. Processing existing records can be useful if, for example, the output topic was accidentally deleted, and you need to recreate the lost data by reprocessing those existing records.
+
+You can specify the exact point from which the transform function should start processing records in the input topic, starting from a specific offset or a specific timestamp. These options ensure that you can handle various scenarios where reprocessing is necessary to maintain data integrity and resilience.
```
>specific offset

There is no option to specify an absolute offset. Records within a given _partition_ are totally ordered, but there's no absolute index for a record within a topic with multiple partitions. So `--from-offset` always specifies an offset delta. The most common usage is probably something like `--from-offset +0`, which means "start from the beginning of each partition".

Lmk if the help text in my other comment is unclear; happy to flesh it out a bit more.

oleiman replied:
>specific timestamp

nitpick (pedantic): first record (per partition of the input topic) committed after the given timestamp. this one is definitely a nitpick.

oleiman commented:
@JakeSCahill looks good, but I think some wires got crossed with respect to the new rpk usage. lmk if I've misdocumented it somewhere, and I'll try to sync everything up.

oleiman reviewed (changes requested):
.

JakeSCahill commented on modules/develop/pages/data-transforms/deploy.adoc:
```diff
@@ -4,6 +4,7 @@
 
 {description}
 
+[[prerequisites]]
```
This is going in beta docs only for now. So the version placeholder will automatically be replaced with the 24.2 RC version

JakeSCahill commented:
Thanks @oleiman ! I was going off the example in the [Jira](https://redpandadata.atlassian.net/browse/CORE-2841) 😅 It should be correct now.
